src/missing_deps.cc

```diff
@@ -0,0 +1,160 @@
+    if (depfile.empty())
+      return;
+    string err;
+    dep_loader.LoadDepfileDeps(edge, depfile, &depfile_deps, &err);
+    if (!depfile_deps.empty())
+      ProcessNodeDeps(node, &depfile_deps[0], depfile_deps.size());
+  }
+}
+
+void MissingDependencyScanner::ProcessNodeDeps(Node* node, Node** dep_nodes,
+                                               int dep_nodes_count) {
+  Edge* edge = node->in_edge();
+  set<Edge*> deplog_edges;
+  for (int i = 0; i < dep_nodes_count; ++i) {
+    Node* deplog_node = dep_nodes[i];
+    // Special exception: A dep on build.ninja can be used to mean "always
+    // rebuild this target when the build is reconfigured", but build.ninja is
+    // often generated by a configuration tool like cmake or gn. The rest of
+    // the build "implicitly" depends on the entire build being reconfigured,
+    // so a missing dep path to build.ninja is not an actual missing dependecy
+    // problem.
+    if (deplog_node->path() == "build.ninja")
+      return;
+    Edge* deplog_edge = deplog_node->in_edge();
+    if (deplog_edge) {
+      deplog_edges.insert(deplog_edge);
+    }
+  }
+  vector<Edge*> missing_deps;
+  for (set<Edge*>::iterator de = deplog_edges.begin(); de != deplog_edges.end();
+       ++de) {
+    if (!PathExistsBetween(*de, edge)) {
+      missing_deps.push_back(*de);
+    }
+  }
+
+  if (!missing_deps.empty()) {
+    set<string> missing_deps_rule_names;
+    for (vector<Edge*>::iterator ne = missing_deps.begin();
+         ne != missing_deps.end(); ++ne) {
+      for (int i = 0; i < dep_nodes_count; ++i) {
+        if (dep_nodes[i]->in_edge() == *ne) {
+          generated_nodes_.insert(dep_nodes[i]);
+          generator_rules_.insert(&(*ne)->rule());
+          missing_deps_rule_names.insert((*ne)->rule().name());
+          delegate_->OnMissingDep(node, dep_nodes[i]->path().c_str(),
+                                  (*ne)->rule());
+        }
+      }
+    }
+    missing_dep_path_count_ += missing_deps_rule_names.size();
+    nodes_missing_deps_.insert(node);
+  }
+}
+
+void MissingDependencyScanner::PrintStats() {
+  printf("Processed %d nodes.\n", static_cast<int>(seen_.size()));
+  if (HadMissingDeps()) {
+    printf("Error: There are %d missing dependency paths.\n",
+           missing_dep_path_count_);
+    printf(
+        "%d targets had depfile dependencies on %d distinct generated inputs "
+        "(from %d rules) without a non-depfile dep path to the generator.\n",
+        static_cast<int>(nodes_missing_deps_.size()),
+        static_cast<int>(generated_nodes_.size()),
+        static_cast<int>(generator_rules_.size()));
+    printf(
+        "There might be build flakiness if any of the targets listed above are "
+        "built alone, or not late enough, in a clean output directory.\n");
+  } else {
+    printf("No missing dependencies on generated files found.\n");
+  }
+}
+
+bool MissingDependencyScanner::PathExistsBetween(Edge* from, Edge* to) {
+  EdgePair key;
```

Comment:
Why not provide a constructor for EdgePair, so you can formulate this as

EdgePair key(from, to);

?